src/common/types.ts

```diff
@@ -95,13 +95,37 @@ function createTypeAliasDeclarationTypeWithSchemaObject(
 
   if (isOneOfOrAnyOfObject(item)) {
     const items = item.oneOf || item.anyOf;
+
+    const reverseDiscriminatorLookup = Object.fromEntries(
+      Object.entries(item.discriminator?.mapping ?? {}).map((x) => x.reverse())
+    );
+
     if (items) {
       return ts.factory.createUnionTypeNode(
-        items.map((oneOrAnyItem) =>
-          isReferenceObject(oneOrAnyItem)
-            ? createTypeRefFromRef(oneOrAnyItem)
-            : createTypeAliasDeclarationTypeWithSchemaObject(oneOrAnyItem)
-        )
+        items.map((oneOrAnyItem) => {
+          if (isReferenceObject(oneOrAnyItem)) {
+            const discriminator = reverseDiscriminatorLookup[oneOrAnyItem.$ref];
+            const discriminatorIntersection = discriminator
+              ? [
+                  ts.factory.createTypeLiteralNode([
+                    ts.factory.createPropertySignature(
+                      undefined,
+                      item.discriminator?.propertyName as string,
```

Comment:
No type assertions please. The `as string` should not be needed, if the types do not match, please create a function guard to assert if it's a string or not